src/NHibernate/Type/ComponentType.cs

```diff
@@ -600,6 +600,11 @@ public override bool IsEqual(object x, object y, ISessionFactoryImplementor fact
 			{
 				return false;
 			}
+
+			var componentType = EntityMode == EntityMode.Poco ? ComponentTuplizer.MappedClass : typeof(IDictionary);
+			if (!componentType.IsInstanceOfType(x) || !componentType.IsInstanceOfType(y))
+				return false;
```

Comment:
`GetPropertyValues` calls below can only fail if the parameter is not of the expected type. So an additional check on the type of object to compare together has been added in the `IsEqual` implementation.